### DISCUSSION TITLE
Memory/Resource management when using the Receiver type function parameter in the spawn call

### DISCUSSION THREAD

**Discussion by yanzacheus** [statelyai/xstate#4585] [Q&A] [answered]
Hello, I'm asking for help to understand exactly how the Receiver type function works in the spawn call:

```
export declare type Receiver<TEvent extends EventObject> = (listener: {
    bivarianceHack(event: TEvent): void;
}['bivarianceHack']) => **void**;
...
spawn(...., receive: Receiver<any>)
```

In fact, how is it handled regarding memory/resource management (what is behind), because the way I use it implies a lot of invocations (see the usage at the bottom of the page, last code ). I just know it returns void.
I will try to give more information about the context and what I am trying to develop with this. 
Perhaps someone knows how to handle it in a better way, or knows other way to do it with version 5 of xstate, or knows if someone already did something like this with xstate and cypress ? 

The web client is an angular application and the backend is composed of several microservices. 
I test the app with cypress
To test it, I mock all the backend mostly for performance, stability, and edge cases (cy.intercept does the job)
That’s why I prefer to define these tests as functional ones and not e2e ones (there are also e2e tests when all microservices are deployed in an environment)
I use Xstate not to manage the state of the app, but to manage the state of « the backend » (ie all the microservices)

Before each test, I create a « backend » machine which computes the resources needed for this test, and consequently the microservice state  machines needed (each micro service manages a number of resources).

```
export const createBackend = (ac: ApplicationContext, resources: BackResource[]) => createMachine<BackendContext, BackendEvent, any>({
  id: Actor.BROWSER,
  context: {
    actorsId: [],
    readyActors : 0,
    apiDomain: getAPIDomain(ac.env),
    tenant: tenantsConfig.get(ac.tenant)
  },
  initial: 'init',
  states: {
    'init': {
      entry: assign({
        actorsId: (context) => [...resources].map((resource) => {
            const service: BackService = resourcesMap.get(resource);
            return spawn(service.builder(context.tenant, context.apiDomain, resources));
          })
      }),
      on: {
        'READY': [{
            target: 'ready',
            cond: 'allServicesAreReady',
          }, {
            actions: 'incReadyActors',
          }
        ],
      }
    },
    'ready': {}
  },
}).withConfig({
  guards: {
    allServicesAreReady: (context) => {
      return context.readyActors === context.actorsId.length - 1;
    }
  },
  actions: {
    incReadyActors: assign({ actorsReady: (context) => context.actorsReady + 1 })
  }
});
```

The builder of the service (in service.builder) knows how to build a micro service machine state using the common **createBackendService** function:

```
export const createBackendService = (
  id: Actor, context: ServiceAgentContext<any>, serviceEvents, tenant: TenantConfig, apiDomain: string, neededResources: BackEntityResource[], initialEntities = DEFAULT_ENTITIES
) => createMachine<ServiceAgentContext<any>, any, any>({
  id,
  context,
  initial: 'init',
  states: {
    'init': {
      entry: assign({
        mockMachine
      }),
      on: {
        'INIT_DONE': {
          target: 'ready'
        },
      }
    },
    'ready': {
      entry: sendParent({ type: 'READY' }),
      on: {
        'NO_COMPUTING': {
          target: 'handling_request',
        },
        'READ_BACK_RESOURCE': {
          target: 'handling_request',
        },
        'REQUEST_SENT': {
          target: 'ready'
        },
        ...serviceEvents
      },
    },
    'handling_request': {
      entry: send((context, event) => ({type: 'BACK_RESOURCE_COMPUTED', value: context }), { to: (context) => context.mockMachine,  }),
      on: {
        'REQUEST_SENT': {
          target: 'ready'
        }
      }
    }
  },
});
```

And here the part of the code using **receive**.
**MockMachine** intercepts all the requests computed
So each time a request is made, a function is called, which creates a promise that will be resolved when the response is computed by the corresponding micro service machine state. 
« computed » just means it’s not just responding a constant, but according to what is sent to the micro service, it defines the new state of the micro service (most of the time the database) and then responds

```
export const mockMachine = (context, event) => spawn((sendBack, receive) => {
  context.neededResources.forEach((resource) => {
    const apis: ResourceAPIDefinitions = APIS.get(resource);
    if (apis) {
      apis.forEach((api) => {
        cy.intercept(
          {
            url: `${context.baseUrl}/${api.endpoint}`,
            https: true,
            method: api.method,
          },
          async function(req) {
            let waiting = true;
            const waitWhenUpdating = new Promise<void>((resolve) => {
              receive(event => {
                if (waiting && event.type === 'BACK_RESOURCE_COMPUTED') {
                  waiting = false;
                  const mockedResponse = (event.value.errors.includes[api.id]) ? { statusCode: 503 } : api.responseBuilder && api.responseBuilder(event.value) || DEFAULT_MOCK;
                  req.reply(mockedResponse);
                  resolve();
                  sendBack({ type: 'REQUEST_SENT'});
                }
              })
            });
            sendBack({ type: api.useStateContext || 'NO_COMPUTING', params: req });
            await waitWhenUpdating;
          }
        );
      })
    }
  });
  sendBack('INIT_DONE');
  return () => {};
});


```

Thank you for any assistance you may provide !

**Comment by davidkpiano** [comment]
I can try to research this for you, but would have to dig back into the magical way we were doing this in XState version 4.

Would it be possible to migrate to XState v5? The way that actors are spawned is much less magical and more explicit/predictable there, also with more controlled memory/resource management.

**Reply by yanzacheus** [reply]
Ok yes, but I don't know anything about version 5. I discovered and learned xstate with a frontend master course (perhaps not up-to-date ?)
What are your advices to migrate and benefit from what is perhaps better done in v5 ? (are there guidelines or tuto) ?
What and how to learn to migrate from v4 to v5 ? 
Will you make another course for v5 on frontendmasters platform ? 
Thanks

**Reply by davidkpiano** [reply]
Here is more information about V5: https://stately.ai/blog/2023-12-01-xstate-v5

I hope to make another course on Frontend Masters, but we want to provide as much material for free as possible.

**Reply by yanzacheus** [reply, marked as answer]
ok I will try to check this link and learn what I have to change.
When migrating if problems still remain, I will open another question or issue.
Thanks
